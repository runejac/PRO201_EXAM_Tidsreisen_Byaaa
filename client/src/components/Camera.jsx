import QrScanner from "qr-scanner";
import { useEffect, useRef, useState } from "react";
import { useNavigate } from "react-router-dom";
import { IoIosArrowBack } from "react-icons/io";
import "../css/cameraBackButton.css";

export default function Camera() {
  const [isScanned, setIsScanned] = useState(false);
  const [data, setData] = useState(null);
  let navigate = useNavigate();
  const videoRef = useRef();
  let qrScanner = null;

  useEffect(() => {
    const videoElem = videoRef.current;

    qrScanner = new QrScanner(
      videoElem,
      (qrcode) => {
        if (qrcode) {
          setIsScanned(!isScanned);
          setData(qrcode.data);
        }
      },
      {
        onDecodeError: (error) => {},
        highlightScanRegion: true,
        maxScansPerSecond: 3,
      }
    );
    qrScanner.start();
    const device_height = window.screen.height;
    const device_width = window.screen.width;

    // set height and width of video
    document.getElementById("qr-video").style.width = device_width + "px";
    document.getElementById("qr-video").style.height = device_height + "px";

    data
      ? (navigate(data, { replace: true }), qrScanner.destroy())
      : console.log(data);

    return () => qrScanner.destroy();
  }, [data]);

  useEffect(() => {
    // need to manually hide scan region after result

    isScanned
      ? ((document.getElementsByClassName(
          "scan-region-highlight-svg"
        )[0].style.display = "none"),
        (document.getElementsByClassName(
          "scan-region-highlight"
        )[0].style.display = "none"))
      : console.log("Nothing scanned yet");

    return () => {
      setIsScanned(false);
    };
  }, [isScanned]);

  return (
    <>
      <div id={"camera-back-button-container"}>
        <IoIosArrowBack
          variant="white"
          onClick={() => navigate("/map")}
          size={35}
        />
      </div>
      <video
        style={{
          backgroundColor: "black",
          position: "a",
          overflow: "hidden",
          whiteSpace: "nowrap",
        }}
        id="qr-video"
        ref={videoRef}
      />
    </>
  );
}
