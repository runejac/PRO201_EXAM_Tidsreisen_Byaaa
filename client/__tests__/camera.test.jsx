/**
 * @jest-environment jsdom
 */

import ReactDOM from "react-dom";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import Camera from "../src/components/Camera.jsx";

describe("Camera", function () {
  it("should render Camera", function () {
    const container = document.createElement("div");

    ReactDOM.render(
      <MemoryRouter>
        <Camera />
      </MemoryRouter>,
      container
    );
    expect(container).toMatchSnapshot();
  });
});
