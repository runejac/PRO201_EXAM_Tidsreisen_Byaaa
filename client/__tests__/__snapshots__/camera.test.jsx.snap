// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera should render Camera 1`] = `
<div>
  <div
    id="camera-back-button-container"
  >
    <svg
      fill="currentColor"
      height="35"
      stroke="currentColor"
      stroke-width="0"
      variant="white"
      viewBox="0 0 512 512"
      width="35"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
      />
    </svg>
  </div>
  <video
    id="qr-video"
    style="background-color: black; position: a; overflow: hidden; white-space: nowrap;"
  />
</div>
`;
